### ISSUE TITLE
Allow marking APIs to be sync as part of network policy

### ISSUE THREAD

**Issue by pkapustin** [beckn/protocol-specifications#144]
# Background

Currently, we often see this repeating pattern in our BPP implementations when processing a request:
* Start a background task
* Return ACK
* Perform a callback to the BAP in a background task

In addition to doing extra work, there are multiple potential points of failure here, and implementing proper error handling becomes a complex task.

# Proposal
We suggest to simplify this where possible by allowing to mark certain BPP APIs as synchronous for HTTP-based implementations as part of the network policy. 
If a BPP API is marked as sync, instead of returning ACK, the response body would contain the data that would otherwise come in a callback.

# Example
A confirm BPP API is marked as sync in a taxi network. A customer confirms a taxi ride order. A BAP issues confirm API, and the BPP uses the sync response to communicate that the order has been placed with all the necessary information. No implementation of on_confirm is required. 

# Summary
We see that allowing to mark BPP APIs to be sync as part of network policy gives the following benefits:
- Simpler implementation
- More robust code
- Better performance

**Comment by beckn-dev** [comment]
I agree, this is a valid issue. And I also agree that the network policy _can_ choose to make the message exchange mechanism synchronous. In such a case, as long as the request body has `context.action` equal to say, `search` and the (synchronous) response body has the `context.action` equal to `on_search`, along with an error object, the acknowledgement can be _implied_ or even _concatenated_ with the message in the synchronous response. While it is technically possible, we need to also think about the practical implications of this change. Not that these are unsolvable problems but we need to acknowledge them and provide recommendations and solutions to address the same. Below are some implications that I see. 

**1. Documentation**: The transaction protocol needs to be _decoupled_ from the transport layer protocol. That means, the documentation of the specification can no longer be in Open API 3.x format as it assumes HTTP by default. Maybe Async API would make more sense. 

**2. Interoperability** :  If two networks are adopting two different exchange mechanisms, how will we ensure interoperability between them. Do we need separate bridging interfaces or can this be baked into the protocol spec itself?

**3. Migration Impact** : Moving from an async model to a synchrnous model may introduce tech debt in existing implementations. Although if this mechanism does become a standard, it will be easier to do this sooner than later when the adoption becomes large. 

Adding @pramodkvarma to comment on the same.

**Comment by pkapustin** [comment]
Yes, all these are valid points of course.

Just wanted to share a couple of possible (and pragmatic) solutions to these.
1. **Documentation**. Yes, if core spec wants to be agnostic of transport protocol, OpenAPI is not ideal. However, I think that OpenAPI can still be a good choice for documenting a network-specific instance of the core spec, if the network is HTTP-based. If such network chooses to implement some BPP APIs synchronously, this can also be clearly expressed using OpenAPI.
2. **Interoperability**. I think it is important to note that two different networks will use two different sets of policies documented by their network-specific specs, and that means that usually they are not going to be automatically interoperable. Marking a BPP API as sync is just one example of a network policy. However, if there is a need for two different networks to work together, we can consider bridging solutions, etc.
3. **Migration impact**. Agree, I think it would be nice to recommend considering this in a new network. But it's not like all existing networks need to change to this model (and it wouldn't be feasible in all cases either). I see it more of a useful simplification for relatively new networks.

**Comment by pkapustin** [comment]
We have successfully implemented this as a pilot in KOMN taxi network.
